### ISSUE TITLE
Question about the code for calculating MSE lib/evaluation.py

### ISSUE THREAD

**Issue by PTAEJOOO** [usail-hkust/t-PatchGNN#14]
Dear author, I am truly grateful for your interesting work. I think there might be an issue with the part of your code(lib/evaluation.py) that calculates the MSE. It seems that your code calculates MSE by two steps, first it computes avg error of each variables and then computes avg error along the variables in line 289~295 of def compute_error(...). However, when calculating in such a two-step manner, it is more accurate to calculate a weighted average rather than simply dividing by the number of valid variables in the current batch with predictions, such as "n_avai_var = torch.count_nonzero(mask_count)", to obtain the correct MSE. Your current MSE calculation code is valid only when all variables have the same value in "mask_count = mask.reshape(-1, n_dim).sum(dim=0)".
Therefore, it seems better to modify "compute_error", "compute_all_losses", "evaluation" functions in lib/evaluation.py as attached below. For now, I have left out the code for calculating MAPE. I would greatly appreciate it if you could take a look at the issue related to this. Thanks :)

#############################################################################################
def compute_error(truth, pred_y, mask, func):
	# pred_y shape [n_batch, n_tp, n_dim]
	# truth shape  [n_bacth, n_tp, n_dim]

	if len(pred_y.shape) == 4: 
		pred_y = pred_y.squeeze(dim=0)

	mask = mask.to(torch.bool)

	if(func == "MSE"):
		error = ((truth - pred_y)[mask])**2
	elif(func == "MAE"):
		error = torch.abs((truth - pred_y)[mask])
	else:
		raise Exception("Error function not specified")

	error_avg = torch.mean(error)

	return error_avg

def compute_all_losses(model, batch_dict, batch_dict_):

	pred_y = model.forecasting(batch_dict["tp_to_predict"], 
		batch_dict["observed_data"], batch_dict["observed_tp"], batch_dict["observed_mask"],
		batch_dict_["observed_data"], batch_dict_["observed_tp"], batch_dict_["observed_mask"])

	# Compute avg error of each variable first, then compute avg error of all variables
	mse = compute_error(batch_dict["data_to_predict"], pred_y, mask = batch_dict["mask_predicted_data"], func="MSE")

	################################
	# mse loss
	loss = mse

	results = {}
	results["loss"] = loss
	results["mse"] = mse.item()

	return results

def evaluation(model, dataloader, dataloader_, n_batches):

	total_results = {}
	total_results["loss"] = 0
	total_results["mse"] = 0
	total_results["mae"] = 0

	count = 0 # total number of prediction points on full batch
	for _ in range(n_batches):
		batch_dict = utils.get_next_batch(dataloader)
		batch_dict_ = utils.get_next_batch(dataloader_)

		pred_y = model.forecasting(batch_dict["tp_to_predict"], 
			batch_dict["observed_data"], batch_dict["observed_tp"], batch_dict["observed_mask"],
			batch_dict_["observed_data"], batch_dict_["observed_tp"], batch_dict_["observed_mask"]) 
		
		mse = compute_error(batch_dict["data_to_predict"], pred_y, mask = batch_dict["mask_predicted_data"], func="MSE")
		mae = compute_error(batch_dict["data_to_predict"], pred_y, mask = batch_dict["mask_predicted_data"], func="MAE")

		mask_count = batch_dict["mask_predicted_data"].sum() # number of prediction points on this batch

		total_results["loss"] += mse * mask_count
		total_results["mse"] += mse * mask_count
		total_results["mae"] += mae * mask_count

		count += mask_count # total number of prediction points on full batch

	total_results["loss"] = total_results["loss"] / count
	total_results["mse"] = total_results["mse"] / count
	total_results["mae"] = total_results["mae"] / count

	for key, var in total_results.items(): 
		if isinstance(var, torch.Tensor):
			var = var.item()
		total_results[key] = var

	return total_results

**Comment by willzhang3** [comment]
We also consider simply calculating the mean of all available values across all variables. However, this causes issues that the variables with densely sampled values would dominate the metric. While variables have different densities of values, these variables should be equally important when evaluating the overall forecasting accuracy. This is why we calculate it in such a two-step manner to ensure that distinct variables with different densities contribute the same to the evaluation results.

**Comment by PTAEJOOO** [comment]
I didn't think of that some variables with densely sampled would dominate the metrics. Now i can see why the two-step manner is necessary. Thank you for your reply!!👍
